rust/src/io/external_manifest/commit.rs

```diff
@@ -0,0 +1,438 @@
+                    object_store
+                        .inner
+                        .rename(&staging, &latest_manifest_path(path))
+                        .await?;
+                }
+
+                Ok(version)
+            }
+            // Dataset not found in the external store, this could be because the dataset did not
+            // use external store for commit before. In this case, we try to read the latest from
+            // the object store
+            None => read_manifest(object_store, &latest_manifest_path(path))
+                .await
+                .map(|m| m.version),
+        }
+    }
+
+    async fn get_version(
+        &self,
+        path: &Path,
+        version: u64,
+        object_store: &ObjectStore,
+    ) -> Result<String> {
+        let manifest_path = manifest_path(path, version);
+        // found in object store
+        if object_store.exists(&manifest_path).await? {
+            return Ok(manifest_path.to_string());
+        }
+
+        // not found in object store, try to get from external store
+        let physical_path = self
+            .external_manifest_store
+            .get(path.as_ref(), version)
+            .await
+            .map_err(|_| Error::NotFound {
+                uri: manifest_path.to_string(),
+                location: location!(),
+            })?;
+
+        object_store
+            .inner
+            .copy(&Path::parse(physical_path)?, &manifest_path)
+            .await?;
+
+        Ok(manifest_path.to_string())
+    }
+
+    async fn commit(
+        &self,
+        manifest: &mut Manifest,
+        indices: Option<Vec<Index>>,
+        path: &Path,
+        object_store: &ObjectStore,
+        manifest_writer: ManifestWriter,
+    ) -> std::result::Result<(), CommitError> {
+        // path we get here is the path to the manifest we want to write
+        // use object_store.base_path.as_ref() for getting the root of the dataset
+
+        // step 1: Write the manifest we want to commit to object store with a temporary name
+        let staging_path = make_staging_manifest_path(path)?;
+        manifest_writer(object_store, manifest, indices, &staging_path).await?;
+
+        // step 2 & 3: Try to commit this version to external store, return err on failure
+        self.external_manifest_store
+            .put_if_not_exists(
+                object_store.base_path().as_ref(),
+                manifest.version,
+                staging_path.to_string(),
+            )
+            .await
+            .map_err(|_| CommitError::CommitConflict {})?;
+
+        // step 4: copy the manifest to the final location
+        object_store.inner.copy(
+            &staging_path,
+            path,
+        ).await.map_err(|e| CommitError::OtherError(
+            Error::IO { message: format!("commit to external store is successful, but could not copy manifest to object store, with error: {}.", e) }
+        ))?;
+
+        Ok(())
```

Comment:
Do we still need to update _latest.manifest here?

Reply:
that is handled by the caller. `crate::dataset::write_manifest_file`